### ISSUE TITLE
Reserve Excess Cluster Capacity

### ISSUE THREAD

**Issue by vlerenc** [gardener/machine-controller-manager#26]
# Story
* As user I want to be able to schedule my workload (if of reasonable size) in my shoot cluster in seconds, even if all nodes of the shoot cluster are busy, so that I am more productive.
* As operator I want to be able to create new shoot cluster control planes in the seed clusters in seconds, even if all nodes of the seed cluster are busy, so that I am more productive.

# Motivation
The [cluster autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) seems to react only when pods can't be scheduled on any of the current nodes due to insufficient resources. That means, it takes minutes after the event until the pod can be scheduled.

# Acceptance Criteria
- [ ] We run the machine deployment with excess capacity (in the given limits), a configurable combination of an absolute excess capacity minimum and a percentage, e.g. (calculation here in machines, but may be alternatively based on CPU/Memory, if that's easier/better): max(1, 5% of the current number of machines)

# Implementation Proposal
This could be done in the machine controller manager, but if collides with the scale-down strategy in the [cluster autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler), consider whether a contribution would be possible or do it in the fork (#25).

# Remarks
Rafael Franzke commented: There is an proposal for allowing exactly the above scenario (reserving cluster resources by cluster-autoscaler) which is planned to be release with Kubernetes v1.8 (according to the 1.8 feature tracking spreadsheet). See this for more information: https://github.com/kubernetes/autoscaler/pull/77/files

# Definition of Done
- [ ] Knowledge is distributed: Have you spread your knowledge in pair programming/code review?
- [ ] Unit Tests are provided: Have you written automated unit tests or added manual NGPTT tickets?
- [ ] Integration Tests are provided: Have you written automated integration tests?
- [ ] Minimum API exposure: If you have added public API, was it really necessary/is it minimal?
- [ ] Operations guide: Have you updated the [operations guide](https://github.wdf.sap.corp/kubernetes/kube-docs/wiki/Operations%20Guide)?
- [ ] User documentation: Have you informed Sylvia Pick about user relevant changes?

**Comment by vlerenc** [comment]
Dear colleagues @hardikdr @prashanth26 @ggaurav10 @amshuman-kr after having adopted the cluster autoscaler, do you have an opinion on how much effort this ticket would be? The proposal https://github.com/kubernetes/autoscaler/pull/56 was replaced with https://github.com/kubernetes/autoscaler/pull/77/files, but both were closed. According to https://github.com/kubernetes/autoscaler/pull/77 the feature/idea didn't make it. Do you know why and can we make it work for us?

Background: Right now we still scale the seeds with a fixed size. With excess capacity we could switch to cluster autoscaler, right @rfranzke? That would lower costs, lower ops, and avoid resource shortage and a potential outage.

**Comment by hardikdr** [comment]
Hi Vedran, 
As you have already mentioned, CA community tried to offer this feature but could not get in, mainly due to inactivity for long period. From higher level approach there was to introduce PlaceHolderPods[FakePods] inside CA which exist only in CA and not really in Api-server. Though there were few optimizations proposed which were expected to be incorporated in the Proposal-which never got in.

Though on a side note, we today discussed the possibility of enabling the MCM in certain ways, that MCM can pass taints from MachineObjects to the NodeObjects, and certain higher level tooling can then maintain [NoSchedule/Execute]taints on X number of **Latest** Machine-objects. - maybe a taint-controller parallel to MachineSet/Deployment. We would try not to affect the current functionality of Set/Deployment. This can be discussed separately in a sync, will avoid confusing here.

Overall I can see, this feature would have a nice use-case in seeds, and if agreed on design - efforts are not significant.

**Comment by amshuman-kr** [comment]
With the approach mentioned by @hardikdr, though it is workable, I see a risk of further diverging from the community.

To mitigate this, I can think of two possible approaches.

1. Propose the MCM-level solution for reserving excess capacity in the cluster-api WG and drive adoption of the proposal into the machine-api spec.
1. Explore other ways of enhancing cluster-autoscaler support reserving excess capacity. Some way other than fake pods. Fake pods seems too much like a hack. This approach may or may not be higher effort.

IMHO, we should at least evaluate approach 2 before freezing on approach 1.

**Comment by vlerenc** [comment]
I agree that "fake pods" sounds pretty much horrible. But (1) could actually be a nice approach. The users could easily specify (per worker pool) the number/percentage of excess capacity, but how would you prevent the cluster autoscaler from scaling it down? Ignore scale-down when the excess capacity constraint is violated? But how would you know? Or keeping it hot (VM on, kubelet off), i.e. not registered at the API server?

**Comment by hardikdr** [comment]
How about also checking the possibility of using extra low-priority pods to reserve space ?
https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/
   - This seems also a recommendation from autoscaling-community.

TL;DR : 
- Pods with the lowest-priority would be pre-empted by k8s in-order to accomodate high-priority pods.

**Comment by amshuman-kr** [comment]
Low-prio pods look like the same kind of hack as fake pods. If this is the kind of thing the autoscaler community is thinking of, I am, frankly, horrified!

Then, I would recommend approach 1 :-)

**Comment by hardikdr** [comment]
Hi Vedran, 
Sorry for late response. 
Answering the question:
`but how would you prevent the cluster autoscaler from scaling it down?`
   - We can put an appropriate annotation on Node-object to prevent autoscaler from scaling it down.
I have prepared a rough proposal with probably approaches: https://github.com/gardener/machine-controller-manager/pull/113

**Comment by vlerenc** [comment]
We discussed that Hardik's proposal with low-priority pods may actually be a real option, especially if that is the recommendation from the autoscaling-community. Also, if this works, it could be integrated into the Gardener Seed Cluster Bootstrapper (no effort for MCM) and would even be better than cpu/mem- or node-level excess capacity reservation as we could then make sure to reserve enough (exact) capacity for additional control planes. Also, pod priorities pod priority and preemption features graduated to beta with v1.11. Hardik will take a look whether this would work smoothly.

**Comment by vlerenc** [comment]
Replaced by gardener/gardener#254 for specific seed cluster excess capacity reservation.
